cmd/common: allow to specify gadget runtime configuration using "-f"

flyth opened the pull request:
This allows specifying a manifest file using the "-f" flag.

Gadget instance specs can look like this:

```yaml
version: 1
kind: instance-spec
imageName: trace_exec
name: yoyo
tags:
  - foo
  - bar
paramValues:
  operator.LocalManager.host: true
---
version: 1
kind: instance-spec
imageName: trace_exec
name: gagaga
tags:
  - second
paramValues:
  operator.oci.wasm.gadget_a: 1212
  operator.oci.ebpf.gadget_a: 1212
  operator.LocalManager.host: true
```

When defining multiple specs inside the file like that, this functionality can only be used in combination with `--detach`.

blanquicet commented:
> Also, documentation would be welcomed! But let's first agree on the naming before proceeding with it.
>
IMO, it's a _must_ for the release. But yeah, let's first agree on the naming ... naming ... naming 🤣

flyth commented on pkg/gadget-config/config.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	res := make([]*Config, 0)
+	cfgEntries := multiYamlRegex.Split(string(cfgBlob), -1)
```
I used what we had since I thought there was no other solution - however, I found a way and applied it now.

flyth commented on cmd/common/oci.go:
```diff
@@ -256,12 +260,77 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+
+		paramValueMap := make(map[string]string)
+
+		if inFile != "" {
```
I extracted parts to a function to increase readability.

flyth commented on cmd/common/oci.go:
```diff
@@ -256,12 +260,77 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+				return fmt.Errorf("multiple gadget runtime configurations found in file %s: this is only supported in combination with --detach", inFile)
+			}
+
+			if isDetach {
```
See above, I extracted parts to a new func to increase readability. Now both branches are similar in length.

flyth commented:
Changed:

File is now called `manifest`. A manifest can host multiple things of which we only evaluate those of kind `instance-spec`(maybe `runtime-spec` is better?) - the kind field is mandatory. Also added a mandatory version field to future-proof.

WDYT?

flyth commented on cmd/common/oci.go:
```diff
@@ -327,9 +377,42 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
 		AddOCIFlags(cmd, operatorParams, skipParams, runtime)
 	}
 
+	cmd.PersistentFlags().StringVarP(&inFile, "file", "f", "", "path to gadget runtime configuration to apply")
```
Probably something like "path to manifest file containing the instance specs".

flyth replied:
The plan was to also allow attaching if only one spec is present - but we can (re-)add that properly later and focus on just starting things from the file for now.

mauriciovasquezbernal commented:
> Changed:
> 
> File is now called `manifest`. A manifest can host multiple things of which we only evaluate those of kind `instance-spec`(maybe `runtime-spec` is better?) - the kind field is mandatory. Also added a mandatory version field to future-proof. This was done to align more with the kubernetes naming as discussed.
> 
> WDYT?

I like it. I think the items inside the manifest should be called "instance-specs".

flyth commented:
> > Changed:
> > File is now called `manifest`. A manifest can host multiple things of which we only evaluate those of kind `instance-spec`(maybe `runtime-spec` is better?) - the kind field is mandatory. Also added a mandatory version field to future-proof. This was done to align more with the kubernetes naming as discussed.
> > WDYT?
> 
> I like it. I think the items inside the manifest should be called "instance-specs".

So `instance-specs` as kind vs `instance-spec`? Kubernetes is only using `spec`.

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
 		timeoutDuration := time.Duration(timeoutSeconds) * time.Second
 
-		imageName := args[0]
+		var imageName string
+		if len(args) > 0 {
+			imageName = args[0]
+		}
+
```
What about if the user runs `... run -f foo.yaml myimage:latest`? I think it should fail.

flyth replied:
I don't know - should it?

I could imagine users having setup filters for namespaces / pods etc in a manifest and then want to run different gadgets using it.

mauriciovasquezbernal replied:
IMO it's too confusing to have this mixed operation mode. What about if there are multiple instances on the manifest? What about if the manifest has an image defined but it doesn't match with the one on the CLI?

blanquicet replied:
IMO, it breaks the balance between flexibility and complexity, so it should fail ... "Keep it simple" 🙂

flyth replied:
I don't know. I don't like artificially preventing users from doing things that would work just fine by intuition (well, at least my intuition 😄)

Currently, everything specified on the command-line would take precedence over what is written in the manifest. Even for multiple instances that would be true.

But I seem to be in the minority seeing your comments + upvotes, hehe.

So, just so that we're on the same page, please confirm:
* using `-f` should prevent specifying an image on CLI
* parameters specified on the CLI should be ignored (or warned about? tricky.) when using `-f`

Careful, though: this would (without introducing exceptions) also apply to connection related settings (like setting the `--remote-address`, for example).

mauriciovasquezbernal replied:
Perhaps I'm misunderstanding something. Is it possible to specify things like `--remote-address`, `--connection-timeout` (and other global parameters) on the manifest file? I would expect this not to be possible as global params **aren't** related to a specific gadget instance. 

For "instance parameters", I think they won't be available as not gadget image is specified.  Hence, in my understanding there is not any parameter that can be overwritten:
- global params: only available on the CLI 
- instance params: only available on the manifest

flyth replied:
You're totally right about the global params, I mixed those up.

Instance params _are_ available, but only if a gadget image is specified - but I will change it to the behavior you described.

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+			detachedParam := runtimeParams.Get("detach")
+			isDetach := detachedParam != nil && detachedParam.AsBool()
+
+			if len(specs) > 1 && !isDetach {
+				return fmt.Errorf("multiple gadget runtime configurations found in file %s: this is only supported in combination with --detach", inFile)
+			}
```
nit: move this after the `if isDetach` conditional? (It seems to me it breaks a little bit the flow of the logic in the current position).

If it's moved below, then it could just be `if len(specs) > 0` as the isDetach branch was handled before.

mauriciovasquezbernal commented on pkg/gadget-manifest/instance-spec.go:
```diff
@@ -0,0 +1,66 @@
+	KindInstanceSpec = "instance-spec"
+)
+
+type InstanceSpec struct {
+	Version     int               `json:"version" yaml:"version"`
+	Kind        string            `json:"kind" yaml:"kind"`
+	ImageName   string            `json:"imageName" yaml:"imageName"`
+	ID          string            `json:"id" yaml:"id"`
+	Name        string            `json:"name" yaml:"name"`
+	Tags        []string          `json:"tags" yaml:"tags"`
+	ParamValues map[string]string `json:"paramValues" yaml:"paramValues"`
+}
```
How should we validate this struct? Should we define a yaml schema?

flyth replied:
What exactly do you mean? Applying stricter rules to catch extra fields for example? Or to provide things like auto-completion in IDEs?

Currently, we check for `version` and `kind` when parsing, and for `image` before running.

mauriciovasquezbernal replied:
I meant to check things like:
- Name can't contain white spaces 
- Tags can't contain `,`
- etc.

flyth replied:
You're right, it makes sense to add those.

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+
+			if isDetach {
+				// Copy special oci params
+				ociParams.CopyToMap(paramValueMap, "operator.oci.")
```
I don't understand this logic. It's copying some parameters to the map, but `ParamValues` on the spec is totally ignored. 

Perhaps something like can help:

```bash 
diff --git a/cmd/common/oci.go b/cmd/common/oci.go
index eb407ce0c..e428af186 100644
--- a/cmd/common/oci.go
+++ b/cmd/common/oci.go
@@ -288,8 +288,8 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn

                        if isDetach {
                                // Copy special oci params
-                               ociParams.CopyToMap(paramValueMap, "operator.oci.")
-                               return runInstanceSpecsDetached(ctx, runtime, specs, runtimeParams, paramValueMap,
+                               //ociParams.CopyToMap(paramValueMap, "operator.oci.")
+                               return runInstanceSpecsDetached(ctx, runtime, specs, runtimeParams, ociParams,
                                        gadgetcontext.WithDataOperators(ops...),
                                        gadgetcontext.WithTimeout(timeoutDuration),
                                        gadgetcontext.WithUseInstance(commandMode == CommandModeAttach),
@@ -387,11 +387,15 @@ func runInstanceSpecsDetached(
        runtime runtime.Runtime,
        specs []*gadgetconfig.InstanceSpec,
        runtimeParams *params.Params,
-       paramValueMap map[string]string,
+       ociParams *params.Params,
        runOptions ...gadgetcontext.Option,
 ) error {
        var merr error
        for _, spec := range specs {
+               // TODO: Clone map?
+               paramValueMap := spec.ParamValues
+               ociParams.CopyToMap(paramValueMap, "operator.oci.")
+
                imageName := spec.ImageName

                // Set some well-known params
```

flyth replied:
Yeah, I broke it with the recent push, sorry :/

Added some tests to make sure it's valid.

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+			runtimeParams.Set("id", cfg.ID)
+			runtimeParams.Set("name", cfg.Name)
+			runtimeParams.Set("tags", strings.Join(cfg.Tags, ","))
+		}
```
Add `paramValueMap = cfg.ParamValues`, otherwise parameters on the file will be ignored when running `... run -f foo.yaml` (without --detach)

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
 			imageName = gadgetInstanceID
 		}
 
+		if imageName == "" {
+			return fmt.Errorf("no image name provided")
+		}
```
When can this happen? only when using `-f` and having the yaml template without the image? IMHO this validation should be in `InstanceSpecsFromReader`.

flyth replied:
Added validation to the spec reader.

mauriciovasquezbernal replied:
So we can remove this check here, right?

- if not using "-f" help will be printed
- if using "-f" the spec reader will complain

mauriciovasquezbernal commented:
> So instance-specs as kind vs instance-spec? Kubernetes is only using spec.

Nope, I meant `instance-spec`s. Kind should be `instance-spec` as it refers to a single one.

mauriciovasquezbernal reviewed (commented):
It looks very good! 


Few general questions:
- Are planning to provide a way to delete the whole manifest? Like `gadgetctl delete -f manifest.yaml` (I suppose it'll only work if the manifests have the name and/or the ID defined)
- Should we print a warning in case the user defines a non existing parameter? If there is a typo in a parameter, and we don't print any warning, it could take al of time for the user to understand what's wrong. (I have been in that position multiple times). (this can be done later on as an improvement as it's not strictly related to this PR)

blanquicet commented on cmd/common/oci.go:
```diff
@@ -256,12 +261,57 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+				return runInstanceSpecsDetached(ctx, runtime, specs, runtimeParams, paramValueMap,
+					gadgetcontext.WithDataOperators(ops...),
+					gadgetcontext.WithTimeout(timeoutDuration),
+					gadgetcontext.WithUseInstance(commandMode == CommandModeAttach),
```
I don't think this line is necessary. It'll never happen.

blanquicet commented on pkg/gadget-manifest/instance-spec.go:
```diff
@@ -0,0 +1,66 @@
+	ID          string            `json:"id" yaml:"id"`
+	Name        string            `json:"name" yaml:"name"`
+	Tags        []string          `json:"tags" yaml:"tags"`
+	ParamValues map[string]string `json:"paramValues" yaml:"paramValues"`
```
You're never setting it. It's also missing in the run command to take this value and send it in the `RunGadget`.

blanquicet commented on cmd/common/oci.go:
```diff
@@ -327,9 +377,42 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
 		AddOCIFlags(cmd, operatorParams, skipParams, runtime)
 	}
 
+	cmd.PersistentFlags().StringVarP(&inFile, "file", "f", "", "path to gadget runtime configuration to apply")
```
This flag is available for `ig`, but it's not working. Mmmm in general, should it be there?

flyth replied:
It was not correctly initializing the operators in that case, I fixed it.

blanquicet commented on pkg/gadget-manifest/instance-spec.go:
```diff
@@ -0,0 +1,66 @@
+			continue
+		}
+		if cfg.Kind != KindInstanceSpec {
+			return nil, fmt.Errorf("expected kind %q in manifest, got kind %q", KindInstanceSpec, cfg.Kind)
```
```suggestion
			return nil, fmt.Errorf("expected kind %q for entry %d in manifest, got kind %q", KindInstanceSpec, c, cfg.Kind)
```

blanquicet reviewed (commented):
> > I like it. I think the items inside the manifest should be called "instance-specs".
> 
> So `instance-specs` as kind vs `instance-spec`? Kubernetes is only using `spec`.
>
IMO, it should be `instance-spec`.

Looks great! I have some general comments:

- `-f` is available in `ig`, should it be there? If yes, it's not working.
- paramValues are not working
- Do you plan to provide a way to delete them all at once?

mqasimsarfraz commented on pkg/gadget-manifest/instance-spec.go:
```diff
@@ -0,0 +1,69 @@
+	ID          string            `json:"id" yaml:"id"`
+	Name        string            `json:"name" yaml:"name"`
+	Tags        []string          `json:"tags" yaml:"tags"`
+	ParamValues map[string]string `json:"paramValues" yaml:"paramValues"`
```
Just to confirm: `map[string]string `  means we don't support:
```
paramValues:
  operator:
    foo: baz
    bar:  baz
```

flyth replied:
Yeah, we don't. Do you think it's not okay to not support that?

mqasimsarfraz replied:
We support it in [configuration file](https://www.inspektor-gadget.io/docs/latest/reference/configuration#configuration-file) so was wondering if we should keep it same here but I don't have too strong opinion on it!

mauriciovasquezbernal replied:
I think supporting only the "dot" notation is better. It's the same notation used by the Golang API and very similar to the CLI (in the CLI we can use shorter names).

mqasimsarfraz reviewed (commented):
Looking goo!

flyth commented on cmd/common/runwithspec_test.go:
```diff
@@ -0,0 +1,240 @@
+				require.NoError(t, err)
+			}
+		}
+		if tt.Mode == testModeInteractiveOnly || tt.Mode == testModeBoth {
+			t.Run(tt.Name, runTest(false))
+		}
+		if tt.Mode == testModeDetachOnly || tt.Mode == testModeBoth {
+			t.Run(tt.Name, runTest(true))
+		}
```
`testModeBoth` (testing for detached and non-detached mode) is used by default if no `*Only` mode is specified (and thus being used by most tests now).

flyth commented on docs/reference/manifests.mdx:
```diff
@@ -0,0 +1,41 @@
+
+Specifying an `image` is also mandatory, while the rest of the fields are optional.
+
+When specifying `paramValues`, please use the fully qualified parameter names provided with their respective
```
Provided a link to the operators docs and an example reference.

flyth commented on docs/reference/manifests.mdx:
```diff
@@ -0,0 +1,123 @@
+
+This example contains two gadget instance specifications that are divided by three dashes (`---`).
+
+The `apiVersion` and `kind` fields define the specification used. They are mandatory and currently we only support
+this version and kind as part of the manifest.
+
+Specifying an `image` is also mandatory, while the rest of the fields are optional.
```
I'm fine with either way. But we need to come to a conclusion :)

/cc @mauriciovasquezbernal

mauriciovasquezbernal replied:
_I think_ `image` is better than `gadget`. In kubernetes, docker, etc. image is used instead of container.

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -250,31 +263,93 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+			if isDetach {
+				// Copy special oci params
+				ociParams.CopyToMap(paramValueMap, "operator.oci.")
+				return runInstanceSpecsDetached(ctx, runtime, specs, runtimeParams, paramValueMap,
```
super minor nit: `paramValueMap` is empty at this point, what about passing `ociParams` to `runInstanceSpecsDetached` and moving `ociParams.CopyToMap(paramValueMap, "operator.oci.")` into it?

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -250,31 +263,93 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+			runtimeParams.Set("name", spec.Name)
+			runtimeParams.Set("tags", strings.Join(spec.Tags, ","))
+
+			tempParams := spec.ParamValues
+			maps.Copy(tempParams, paramValueMap)
+			paramValueMap = tempParams
```
I have read these lines multiple times and I can't understand what they do. 
Can't it just be `maps.Copy(paramValueMap, spec.ParamValues)`?

mauriciovasquezbernal commented on cmd/common/oci.go:
```diff
@@ -330,6 +406,39 @@ func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumn
+
+		maps.Copy(spec.ParamValues, paramValueMap)
+
+		gadgetCtx := gadgetcontext.New(
+			ctx,
+			image,
+			runOptions...,
+		)
```
`gadgetCtx := gadgetcontext.New(ctx, image, runOptions...)`?

(I kind of feel bad commenting these things, but on the other hand I want to keep it as clean as possible)

flyth replied:
Don't feel bad about it.
I think I left it that way because it felt "more important" that way (the gadgetcontext being the central point of things), but I agree this should be shortened.

mauriciovasquezbernal commented on docs/reference/manifests.mdx:
```diff
@@ -0,0 +1,123 @@
+
+If a manifest contains multiple instance specs, it can only be run by additionally using `--detach`. This is however
+only available when using `gadgetctl`. The run command will then try to create instances for all given specs and return
+either their IDs.
```
```suggestion
their IDs.
```

Or what's the role of "either" in his sentence?

flyth replied:
It was "either an error or..." at one point, but I decided against using it. Didn't catch that when re-reading 😄

mauriciovasquezbernal commented on docs/reference/manifests.mdx:
```diff
@@ -0,0 +1,123 @@
+When running interactively, you are limited to one gadget instance spec per manifest. This applies to `ig` as well as
+`gadgetctl` without the `--detach` flag.
+
+<Tabs groupId="env">
```
The idea of using tabs is to differentiate different ways of performing the same task, for instance, to compare how to run the same gadget with ig, gadgectl, kubectl-gadge, etc.. I think in this case the gadget.yaml should be shown before and  then there can be tabs for ig and gadgetctl.

flyth replied:
Hmm, I've also seen tabs used to show the contents of different files, for example when showing HTML/JS/CSS. But this is kind of mixed. I still think it's useful this way because it shows that the example consists of both the command and the file, without taking more space (we already got an example at the top.)

mauriciovasquezbernal commented on cmd/common/runwithspec_test.go:
```diff
@@ -0,0 +1,260 @@
+	if r.testSpec.ExpectedTags != nil {
+		assert.Equal(r.t, r.testSpec.ExpectedTags, runtimeParams.Get(grpcruntime.ParamTags).AsStringSlice())
+	}
+	if r.testSpec.ExpectedParams != nil {
```
nit: this check is not needed as range over nil works fine.

mauriciovasquezbernal commented on cmd/common/runwithspec_test.go:
```diff
@@ -0,0 +1,260 @@
+		for k, v := range r.testSpec.ExpectedParams {
+			val, ok := paramValueMap[k]
+			require.True(r.t, ok, "expected paramValue exists for %q", k)
+			require.True(r.t, ok, "expected paramValue to be %q, got %q", v, val)
```
`require.Equal(r.t, v, val)` ? 

Also the require package already provides the "expected .. got" message, so we don't need to add it.

mauriciovasquezbernal commented on cmd/common/runwithspec_test.go:
```diff
@@ -0,0 +1,260 @@
+
+	for _, tt := range tests {
+		runTest := func(detach bool) func(t *testing.T) {
+			return func(t *testing.T) {
```
I added `t.Parallel()` and everything exploded :exploding_head:

flyth replied:
It seems that `setFlagsFromConfig()` has global side-effects that cause that (concurrent map read/writes). I think it's fine for the tests.

mauriciovasquezbernal reviewed (approved):
Some nits but LGTM already. Thanks for the work on this!

flyth commented on cmd/common/runwithspec_test.go:
```diff
@@ -0,0 +1,260 @@
+			Manifest: `
+apiVersion: 1
+kind: instance-spec
+image: trace_exec
```
No, it's supposed to throw an error - e.g. the image from the CLI should _not_ be used if a file is specified.

flyth commented:
Thanks for the reviews!
